akka-stream/src/main/scala/DocumentStage.scala

```diff
@@ -65,22 +72,25 @@ private[akkastream] class DocumentStage[T](
         }
       }
 
-      private def killLast(): Unit = last.foreach {
+      private def killLast(): Try[akka.Done] = last.fold[Try[akka.Done]](Success[akka.Done](akka.Done)) {
         case (r, _, _) =>
           kill(r)
       }
 
       @SuppressWarnings(Array("CatchException"))
-      private def kill(r: Response): Unit = {
-        try {
+      private def kill(r: Response): Try[akka.Done] = {
```

Comment:
Rather keep the `try` version: there is no composition there, so no value to compensate the instance creation

Reply:
that `Try` is returned by `killLast` which in tun is composed in `postStop`.

Reply:
The `Promise` can be directly completed for me

Reply:
I don't think so because `kill`/`killLast` is also called from `onFailure` where 
 - if err handler decide to continue we should not complete the promise.
 - if err handler decide to stop (`Fail`/`Done`) we should use the original `reason` exception to complete the promise (but that will happen only in `postStop`)

Reply:
Even if not directly in `killX`, the can/should be able in any case to complete the promise directly according the stage completion.

Reply:
well, `postStop` in the callback that is always called at the stage completion so it looks like the best place to complete the promise.
Where else would you complete the promise at a point that correspond to the stage completion?

Reply:
`postStop` is always called, without knowing why (success or failed), each time `completeStage` or `fail` is called, the promise should also be completed appriopriately

Reply:
I see now,
The reason I was confused is that I believed that `fail` did not actually triggered the stop of the stage in every case, but only if the `SupervisionStrategy` used by the materializer decided to stop the stream on errors... but that does not seem to be the case.

Nevertheless I think, since the stage has only one connection, it would be clearer to call `failStage` instead of `fail`, do you agree?

Reply:
In current case it doesn't really matter